### ISSUE TITLE
[Proposal]Develop a scalable Finetune class to fine tune the paramters of deep network

### ISSUE THREAD

**Issue by stereomatchingkiss** [mlpack/mlpack#458]
I would like to develop a scalable Finetune class, need some suggestions, following are the suggestion api

```
/**
 * Fine tune deep network like StackAutoencoder
 *     
 *@tparam OutputLayerType types of the output layer, must implement three functions.
 * - Gradient(const arma::mat& parameters, arma::mat& gradient);
 * - double Evaluate(const arma::mat& parameters);
 * - arma::mat& GetInitialPoint();
 *@tparam FineTuneGradient Functor for calculating the last gradient, it should 
      implement two functions
 * - template<typename T> Gradient(arma::mat const&, arma::mat const&, T const&, arma::mat&)
 * - Deriv(arma::mat const&, arma::mat&);
 */
template<typename OutputLayerType,
         typename FineTuneGradient>
class FineTuneFunction
{
public:                
    /**
     * Construct the class with given data
     * @param input The input data of the LayerTypes and OutputLayerType
     * @param parameters The parameters of the LayerTypes and OutputLayerType        
     * @param outLayerType The type of the last layer(ex : softmax)
     */
    FineTuneFunction(std::vector<arma::mat*> &input,
                                 std::vector<arma::mat*>  &parameters,                        
                     OutputLayerType &outLayerType);            

    /**
     * Evaluates the objective function of the networks using the
     * given parameters.
     * @param parameters Current values of the model parameters.
     */
    double Evaluate(const arma::mat& parameters);       

    /**
     * Evaluates the gradient values of the objective function given the current
     * set of parameters. The function performs a feedforward pass and computes
     * the error in reconstructing the data points. It then uses the
     * backpropagation algorithm to compute the gradient values.
     * @param parameters Current values of the model parameters.
     * @param gradient Matrix where gradient values will be stored.
     */
    void Gradient(const arma::mat& parameters, arma::mat& gradient);

    //! Return the initial point for the optimization.
    arma::mat& GetInitialPoint();
};
```

The example of using this class(omit the initialization part of the training data)

```
using namespace mlpack;

arma::mat sae1_input ;
arma::mat sae2_input ;   
arma::mat sm_input;
arma::Row<size_t> labels(2);
labels(0) = 0;
labels(1) = 1;
regression::SoftmaxRegressionFunction sm(sm_input, labels, 2);

arma::mat sae1_params;
arma::mat sae2_params;
arma::mat sm_params;

 //after training, the class will update the params
  std::vector<arma::mat*> params{
      &sae1_params, &sae2_params, &sm_params
  };
 //the class will change the input(except the first input) when training
 std::vector<arma::mat*> inputs{&sae1_input, &sae2_input, &sm_input};
  FineTuneFunction<              
          decltype(sm),
          SoftmaxFineTune
          > finetune(inputs, params, sm);

  //create lbfgs to fine tune the value
```

Besides, the SoftmaxFunction need to cache the probabilities value, else you have to recalculate the probabilites two more times when fine tune the parameters, any suggestions?Thanks

**Comment by rcurtin** [comment]
I like that this uses the existing optimizer API, but a couple comments:

I think that `std::vector<arma::mat*>` is a little bit awkward; could you possibly instead take the instantiated network as the input and extract/modify the parameters from there?

What does the function `Deriv(arma::mat const&, arma::mat&)` do?  How is that different from `Gradient()`?

> Besides, the SoftmaxFunction need to cache the probabilities value, else you have to recalculate the probabilites two more times when fine tune the parameters, any suggestions?

This is a common problem in the various optimizers; often an optimizer will make a call to `Evaluate()` followed directly by a call to `Gradient()` with the same parameters.  I don't think this can be fixed unless we modify the API for the optimizers.  One way to do this might be to add an `EvaluateWithGradient()` function that both calculates the objective and also calculates the gradient.  It's relatively straightforward to use template metaprogramming inside of the optimizers in order to call either `Evaluate()` then `Gradient()` or `EvaluateWithGradient()` when available.  So this way, the `EvaluateWithGradient()` function would be optional, but when supplied it could accelerate the computation.  What do you think of this idea?

I can't comment on how well this proposal fits with the rest of the neural network code; I don't know if there are already any plans for fine tuning.

**Comment by stereomatchingkiss** [comment]
>I think that std::vector<arma::mat*> is a little bit awkward; could you possibly instead take the >instantiated network as the input and extract/modify the parameters from there?

There are some problems here
1 : Neither softmax nor stackautoencoder can access the input data, but when you are finetuning the parameters, the input need to update frequently. Pass by pointer is a less intrusive solution I can think of by now(not reference because reference cannot take the reference of reference). We have to let the users access the input data if we want to pass in the instantiated network.

2 : The class softmaxRegression cannot  access parameters. However, about the parameters, pass by std::vector<arma::mat const*> should be better, because I do not know when should I update the parameter, this should be done by another api after the whole training process finished.

3 : Pass by the instantiated network would make the implementation details a little bit complex(need to deal with compile time for loop, that means, we need TMP at here).

ex : 
    auto networkTuple = std::forward_as_tuple(sae1, sae2);
    FineTuneFunction<
    decltype(networkTuple),
    SoftmaxRegressionFunction,
    SoftmaxFineTune
    > finetune(networkTuple, softmax);

>What does the function Deriv(arma::mat const&, arma::mat&) do? How is that different from Gradient()?

The details is listed by UFLDL(http://deeplearning.stanford.edu/wiki/index.php/Fine-tuning_Stacked_AEs)

Gradient() is the derivative related to the last layer of the finetune network(J of number 2)
Deriv() will calculate the derivate of back propagation of neural network(f'(z) of number 3)

Example : 

    class SoftmaxFineTune
    {
     public:
        template<typename T>
        static void Gradient(arma::mat const &input,
                                       arma::mat const &weights,
                                       T const &model,
                                       arma::mat &gradient)
        {
            gradient = ((weights.t() * model.Probabilities()) %
                             (input % (1 - input))) / input.n_cols;
        }

        static void Deriv(arma::mat const &input, arma::mat &output)
        {
            output = input % (1 - input);
        }
    };

If the last layer always should be softmax, then we could remove this template parameter.

>So this way, the EvaluateWithGradient() function would be optional, but when supplied it could 
>accelerate the computation. What do you think of this idea?

This could save one calculation, but you still need to recalculate the probabilities if you do not cache it when finetune(if you are using softmaxRegression as last layer), one calculation is done by Evaluate, one is Gradient, another one is finetune(when finding the last derivative term). Could we just let the SoftmaxRegression cache the probabilities and let the users access it(read only)?

    SoftmaxRegressionFunction
    {
     public:
      //! Gets the probabilities.
      const arma::mat& Probabilities() const { return probabilities; }
     private:
      //! Probability matrix
      arma::mat probabilities;
    };

When you call the Evaluate function, rather than create a temporary variable, use the data member
    probabilities. 

When you call the Gradient function, do not recalculate, just reuse the calculated probabilities, this probabilities only depend on the training data and the weights(parameters), and the weights(parameters) would not update before you call the Gradient function, so it should be safe to cache it(unless there are some optimization algorithms would not call Evaluate before calling Gradient) I try and run the unit test, it works fine.

>I don't know if there are already any plans for fine tuning.

I hope there are, this is a quite useful tool.

Thanks for your helps:).

**Comment by stereomatchingkiss** [comment]
The test of softmax_regression do work after the alternation, but the test case "SoftmaxRegressionFunctionGradient" must add two more lines(at line 131 and 132)

    // Get gradients for the current parameters.
    arma::mat gradient1, gradient2;
    //add these two function calls to evaluate the objective value first, else the probabilites matrix
    //would not build
    srf1.Evaluate(parameters);
    srf2.Evaluate(parameters);

**Comment by stereomatchingkiss** [comment]
I do some change on the api of FineTuneGradient, this api may provide better optimization codes

    class SoftmaxFineTune
    {
     public:
      template<typename T>
      static void LastGradient(arma::mat const &input,
                               arma::mat const &weights,
                               T const &model,
                               arma::mat &gradient)
        {
            gradient = (weights.t() * model.Probabilities()) /
                    static_cast<double>(input.n_cols);
            gradient = gradient % (input % (1 - input));
        }

        static void Gradient(arma::mat const &input,
                             arma::mat const &weight,
                             arma::mat const &deriv,
                             arma::mat &output)
        {
            output = (weight.t() * deriv) % (input % (1 - input));
        }
    };

**Comment by rcurtin** [comment]
I think that the codebase has changed so much that unfortunately this proposal may not work anymore.  I don't have any problem with the idea, but in the interest of keeping the issue tracker manageable I'll close this for inactivity.  If anyone else wants to take up the issue we can reopen it.
